### DISCUSSION TITLE
Build the web site using makdown direct renderer and not static site generators

### DISCUSSION THREAD

**Discussion by mckaygerhard** [BashGui/easybashgui#28] [Ideas]
Raito https://raito.arnaud.at/ is a dynamic website renderer, **unlike frameworks as hugo, yekil(github) and others that must generate the html pages first to then be placed on the website, Raito generates them on the fly, just when the user requests the specifies page to display**. This is done by taking the markdown files and making HTML of each them when each one are requested in a request.

**Why not jekil (github) deploys??**, the proposal of https://github.com/BashGui/easybashgui/issues/20 is complicated, it require a understanding of jeckil deploy on each change of theme, raito only require CSS changes, github deploy with yekil instead requires not only CSS changes also requires YAML knowledge and changes.. 

## Deploy of the web site

There already others ways more traditional, and easy to understand: https://github.com/codeling/w2wiki that is in the same way, despiste a fork, this m2wiki allows to sync a git repository so indeed we can talk that it makes the same as raito, but in reverse way, people edit at the site and the changes will be committed to the repo, unfortunately this does not handle authentication neither support real git sync (only comit and push).

I already made a git web site deploy, at the branch https://github.com/venenux/easybashgui/tree/webpage-with-raito , but using raito.. 

* becouse if w2wiki does not support authentication.. makes no sense usage of w2wiki
* so raito is simple and easy, the changes are directly from git repository, with w2wiki you must sync and setup

The web page can be tested from the github pages.. for the branch i tested its from https://venenux.github.io/easybashgui/ the process can be also adapted to use a customized domain (by example https://sites.google.com/site/easybashgui/ cannot be used but https://sites.google.com can be used.. so we must check the way to define root web deploy ..)

### how to use raito with google sites

Google site old way is what i tested long time ago, so easy, as i remembered, just take the 3 markdown files of the repository (with pictures also of course) in same structure and also the index.html raito file and put on your google site, later edit the index.html and define the root deploy with "/site/easybashgui/" and that's all!

1. take one "index.html" raito file and customize the CSS themes, dont fogott to put the root in index. html with  "/site/easybashgui/"
2. the markdown files are already on the github repository .. no php need, no cgi need! take it with the index.html raito file and put them all into the root of the google sites web server (if you can upload files in raw, just create ir as files to download)
3. back to main repo, at the right side of "about" pust the icon config and there below pust the url of the google site

**How this work?**: it **used the natural web server of google sites pages**, all the work of rendering will be made by raito, not by webserver that will only renders the index.html one time and rest of work will be done by the browser client.. **the index.html generate the page from the README.md file on the fly**.. if you change it the changes are intermediately at the deploy.. no dependency of the yekil engine of the github crap! cos i hate M$ company..

**What about the repository? and automatic update of the site** this its not possible if you use google site, but is not a problem, mostly updates to those files are not so often, but of course in my personal opinion i prefers the github way, cos its full automatize!

### how to use the github pages feature instead the google sites

The process is in fact more easy cos will be automatized:

1. put the index.html into root of your dessired (modified of course) of raito!
2. the markdown files are already on the github repository .. no php need, no cgi need! none engine need! all rendering process is made at the client side
3. go to git repo github web interface oif EBG, go to settings tab, search for pages at the left site
4. enable by selecting the branch "master" and then selecting in second intpu the "root" (will take the README.md and docs/ )
5. back to main repo, at the right side of "about" pust the icon config and there below select "Use your GitHub Pages website"

After save, the web page URL will be at the about of the repository, wait 5 minutes and web page will be displayed

**How this work?**: it **used the natural web server of github pages BUT do not use any deploy generation of github**, that wil break if you dont know about jekil theme deploy and depends on python crap software.. all the work of rendering will be made by raito, not by webserver that wil only renders the index.html one time and rest of work will be done by the browser client **the index.html generate the page from the README.md file on the fly**.. if you change it the changes are inmmediatelly at the deploy.. no dependency of the yekil engine of the github crap! cos i hate M$ company..

**Comment by vaisarger** [comment]
Hi all !
First of all, let me thank @mckaygerhard for this detailed explanation about all options for having a dynamic documentation page of a certain project.

Some years ago, I had many servers with different GNU/Linux distributions and thus different sets of packages installed... since I needed to find a solution to give a graphical interface to all my scripts (without every time rewrite them depending on which widget was present in each server ), I started writing first core of EBG *but* always with **ease** in my mind.

So, even addressing this issue, I like an easy (very easy ) solution:
I would put a simple link in my previous site to a dynamic page describing EBG, like the link I actually put:
https://sites.google.com/site/easybashgui
where, in fact, now there is (in the drop down menu ) the link "readme & docs" (temporarily pointing to https://venenux.github.io/easybashgui/#/ , but I'll change it as EBG index.html file would be ready in its Github page ).

In this way, I'm going to keep some old historical pages but especially some useful pages like "easybashgui functions", the latter useful for me too to review the synopsis of some functions for which even I had forgotten some details (e.g.: tagged_menu() wants before the tag and then the item, or the item and then the tag? _- I confess that I still don't remember it , ha ha... -_

Anyway... I prefere this solution, if you don't mind... ;)

**Reply by mckaygerhard** [reply]
oh yes.. i must test ALL those functions.. `tagged_menu()` where soso interesting.. :laughing: 

just to be sure: you just can made you own copy repo and put the index.html file there! you are not forced to use the venenux fork..

**Reply by vaisarger** [reply]
I know... in fact, I was waiting for you (ehr... ) to add a new shiny "index.html" in master EBG Github... :D

**Reply by mckaygerhard** [reply]
:face_exhaling:  oh the work was delegated :laughing:  puff !!! that's how all businessmen are

ok easyle.. but still a hint! the CSS i made in the venenux branch is just a trick, in fact actually I just put a basic menu without much effort and linked the boostrap theme... I didn't even choose a good higlight.css theme for the source code parts...

what do I mean? After I upload it, if you wish, you can alter it to improve its appearance or have others do it. @vaisarger  so maybe now with that @zen0bit will loook into and liked to tune it!

**Reply by mckaygerhard** [reply]
oh let me include the web download .. as you pointed just now

**Reply by vaisarger** [reply]
😆😆😆😆
Well... I'm a poor businessman... 😆😆😆

Out of joke... Of course my project is always open to contribution of all kinds... _I'm not good at all in graphic_, and all contributions are welcome (  @zen0bit , if you'd like tune it, you are surely welcome! ).

**Reply by mckaygerhard** [reply]
done: https://github.com/BashGui/easybashgui/pull/29 with a download link.. i am searchig for badgets on the link of OBS.. just one commit more! and in 15 minutes is done

**Comment by vaisarger** [comment]
Thanks to great @mckaygerhard , even a dynamic page related to EBG project is available at:
https://bashgui.github.io/easybashgui/#/

Thank you very much !
